src-self-hosted/Module.zig

```diff
@@ -1309,7 +1309,7 @@ fn astGenAndAnalyzeDecl(self: *Module, decl: *Decl) !bool {
                     !gen_scope.instructions.items[gen_scope.instructions.items.len - 1].tag.isNoReturn()))
                 {
                     const src = tree.token_locs[body_block.rbrace].start;
-                    _ = try astgen.addZIRNoOp(self, &gen_scope.base, src, .returnvoid);
+                    _ = try astgen.addZIRNoOp(self, &gen_scope.base, src, .@"return");
```

Comment:
This is missing a "set return value to void" instruction, which would be necessary for the "expected T, found void" compile error (when one forgets to return a value). I think we should keep the existing `returnvoid` instruction as well as the "return with an operand" instruction, and what this branch can do is add an additional "set the return value" instruction and "return control flow assuming the return value has already been set".

Idea being that we can have IR instructions with overlapping responsibilities; it's easy to have them call the appropriate functions during semantic analysis. This is in the effort of lowering memory usage; adding another enum tag to `ir.Inst.Tag` is free, so we may as well minimize the data we are allocating for these instructions.

**However** another observation is that once we do defers, it would make sense to have one canonical "exit the function" path from any scope. I think we should actually *remove* the return control flow instruction.

So, here's my proposal:

 * Add `set_ret_val` instruction, which does no control flow.
 * Add `set_ret_void` instruction. Same as `set_ret_val` but the operand is assumed to be the void value.
 * Remove `return` and `returnvoid` instructions.

There would be no more return control flow logic. It would be implied at the end of the main outer block of a function, and early-return control flow would be managed with breaks. I think this would work well with how to codegen defer expressions. Note, however, that the return value would either be set with `ret_ptr` and writing through the pointer, or `set_ret_val` directly.

The astgen for `return` syntax will need to change, to use a combination of the "break" and "set_ret_val" instructions rather than emitting the now-deleted "return" control flow instruction.